File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/busca/BuscaProfundidade.java
package busca;

import java.util.LinkedList;
import java.util.List;


/**
 *   Algoritmos de Busca (geral, qquer problema)
 *   Busca a solução por busca em profundidade.
 *
 *   @author Jomi Fred Hübner
 */
public class BuscaProfundidade extends Busca {

    protected int profMax = 40;

    /** busca sem mostrar status */
    public BuscaProfundidade() {
    }
    public BuscaProfundidade(int m) {
        profMax = m;
    }
    
    /** busca mostrando status */
    public BuscaProfundidade(MostraStatusConsole ms) {
        super(ms);
    }
    public BuscaProfundidade(int m, MostraStatusConsole ms) {
        super(ms);
        profMax = m;
    }
    
	
	public void setProfMax(int m) {
		profMax = m;
	}
	
    public Nodo busca(Estado inicial) {
        status.inicia();
        initFechados();
        
        List<Nodo> abertos = new LinkedList<Nodo>();
        
        abertos.add(new Nodo(inicial, null));
        
        while (!parar && abertos.size() > 0) {
            
            Nodo n = abertos.remove(0);
            status.explorando(n,abertos.size());
            if (n.estado.ehMeta()) {
                status.termina(true);
                return n;
            }
        
            if (n.getProfundidade() < profMax) {
                abertos.addAll( 0, sucessores(n) );
            }            
        }
        status.termina(false);
        return null;
    }        
    
    public String toString() {
    	return "BP - Busca em Profundidade";
    }
}

File: 1 - metodosBusca/python/app/core/busca/busca_profundidade.py
from collections import deque
from .busca import Busca
from .nodo import Nodo
from .status import Status

class BuscaProfundidade(Busca):
    def __init__(self, prof_max=40, mostra_status_console=None):
        super().__init__(mostra_status_console)
        self.prof_max = prof_max

    def set_prof_max(self, m):
        self.prof_max = m

    def busca(self, inicial):
        self.status.inicia()
        self.init_fechados()
        
        abertos = deque([Nodo(inicial, None)])
        
        while not self.parar and abertos:
            n = abertos.popleft()
            self.status.explorando(n, len(abertos))
            if n.estado.eh_meta():
                self.status.termina(True)
                return n
        
            if n.get_profundidade() < self.prof_max:
                abertos.extendleft(reversed(self.sucessores(n)))
                
        self.status.termina(False)
        return None

    def __str__(self):
        return "BP - Busca em Profundidade"